File: 01_core_backend/src/gateway/middleware/service-auth.js
/**
 * Service Authentication Middleware
 * JWT-based authentication for service-to-service communication
 */

const jwt = require('jsonwebtoken');
const crypto = require('crypto');

class ServiceAuthenticator {
  constructor() {
    this.serviceSecret = process.env.SERVICE_JWT_SECRET || this.generateDefaultSecret();
    this.serviceName = process.env.SERVICE_NAME || 'api-gateway';
    this.tokenExpiry = process.env.SERVICE_TOKEN_EXPIRY || '15m';
    this.enabled = process.env.SERVICE_AUTH_ENABLED !== 'false';
    
    // Cache for active tokens to avoid regenerating
    this.tokenCache = new Map();
    this.cacheTimeout = 5 * 60 * 1000; // 5 minutes
    
    console.log(`🔐 Service Authentication: ${this.enabled ? 'ENABLED' : 'DISABLED'}`);
  }

  generateDefaultSecret() {
    // Generate a secure random secret for development
    const secret = crypto.randomBytes(64).toString('hex');
    console.warn('⚠️ Using generated JWT secret. Set SERVICE_JWT_SECRET in production!');
    return secret;
  }

  /**
   * Generate a service JWT token for API calls
   */
  generateServiceToken(targetService, payload = {}) {
    if (!this.enabled) {
      return null;
    }

    const cacheKey = `${targetService}:${JSON.stringify(payload)}`;
    const cached = this.tokenCache.get(cacheKey);
    
    // Return cached token if still valid
    if (cached && Date.now() < cached.expires) {
      return cached.token;
    }

    const tokenPayload = {
      iss: this.serviceName, // Issuer
      aud: targetService,     // Audience
      sub: 'service-auth',    // Subject
      iat: Math.floor(Date.now() / 1000),
      jti: crypto.randomUUID(), // Unique token ID
      ...payload
    };

    const token = jwt.sign(tokenPayload, this.serviceSecret, {
      expiresIn: this.tokenExpiry,
      algorithm: 'HS256'
    });

    // Cache the token
    const expiresAt = Date.now() + (this.getExpirySeconds() * 1000) - 30000; // 30s buffer
    this.tokenCache.set(cacheKey, {
      token,
      expires: expiresAt
    });

    return token;
  }

  /**
   * Validate incoming service JWT token
   */
  validateServiceToken(token, expectedAudience = null) {
    if (!this.enabled) {
      return { valid: true, bypass: true };
    }

    if (!token) {
      return { 
        valid: false, 
        error: 'No service token provided',
        code: 'MISSING_TOKEN'
      };
    }

    try {
      const decoded = jwt.verify(token, this.serviceSecret, {
        algorithms: ['HS256']
      });

      // Validate audience if specified
      if (expectedAudience && decoded.aud !== expectedAudience) {
        return {
          valid: false,
          error: `Invalid token audience. Expected: ${expectedAudience}, Got: ${decoded.aud}`,
          code: 'INVALID_AUDIENCE'
        };
      }

      // Validate issuer is from known services
      const validIssuers = ['api-gateway', 'ai-engine', 'vision-service'];
      if (!validIssuers.includes(decoded.iss)) {
        return {
          valid: false,
          error: `Unknown service issuer: ${decoded.iss}`,
          code: 'UNKNOWN_ISSUER'
        };
      }

      return {
        valid: true,
        payload: decoded,
        issuer: decoded.iss,
        audience: decoded.aud
      };

    } catch (error) {
      let errorCode = 'TOKEN_VALIDATION_ERROR';
      let errorMessage = error.message;

      if (error.name === 'TokenExpiredError') {
        errorCode = 'TOKEN_EXPIRED';
        errorMessage = 'Service token has expired';
      } else if (error.name === 'JsonWebTokenError') {
        errorCode = 'INVALID_TOKEN';
        errorMessage = 'Invalid service token format';
      }

      return {
        valid: false,
        error: errorMessage,
        code: errorCode
      };
    }
  }

  /**
   * Middleware for validating incoming service requests
   */
  createValidationMiddleware(expectedAudience = null) {
    return async (request, reply) => {
      if (!this.enabled) {
        // Add bypass flag for logging
        request.serviceAuth = { bypass: true };
        return;
      }

      const token = this.extractToken(request);
      const validation = this.validateServiceToken(token, expectedAudience);

      if (!validation.valid) {
        return reply.status(401).send({
          error: 'Service Authentication Failed',
          message: validation.error,
          code: validation.code,
          timestamp: new Date().toISOString()
        });
      }

      // Attach service info to request
      request.serviceAuth = {
        valid: true,
        issuer: validation.issuer,
        audience: validation.audience,
        payload: validation.payload
      };
    };
  }

  /**
   * Extract token from request headers
   */
  extractToken(request) {
    const authHeader = request.headers.authorization;
    
    if (authHeader && authHeader.startsWith('Bearer ')) {
      return authHeader.substring(7);
    }
    
    // Also check custom service header
    return request.headers['x-service-token'];
  }

  /**
   * Add service authentication headers to outgoing requests
   */
  addServiceHeaders(targetService, existingHeaders = {}) {
    if (!this.enabled) {
      return existingHeaders;
    }

    const token = this.generateServiceToken(targetService);
    
    return {
      ...existingHeaders,
      'Authorization': `Bearer ${token}`,
      'X-Service-Token': token,
      'X-Service-Name': this.serviceName,
      'X-Request-ID': crypto.randomUUID()
    };
  }

  /**
   * Get token expiry in seconds
   */
  getExpirySeconds() {
    const expiry = this.tokenExpiry;
    if (typeof expiry === 'number') return expiry;
    
    // Parse string format like '15m', '1h', '30s'
    const match = expiry.match(/^(\\d+)([smhd])$/);
    if (!match) return 900; // Default 15 minutes
    
    const value = parseInt(match[1]);
    const unit = match[2];
    
    switch (unit) {
      case 's': return value;
      case 'm': return value * 60;
      case 'h': return value * 3600;
      case 'd': return value * 86400;
      default: return 900;
    }
  }

  /**
   * Clear token cache (useful for testing)
   */
  clearTokenCache() {
    this.tokenCache.clear();
  }

  /**
   * Get authentication status
   */
  getStatus() {
    return {
      enabled: this.enabled,
      serviceName: this.serviceName,
      tokenExpiry: this.tokenExpiry,
      cachedTokens: this.tokenCache.size,
      hasSecret: !!this.serviceSecret
    };
  }
}

// Export singleton instance
const serviceAuth = new ServiceAuthenticator();

module.exports = {
  ServiceAuthenticator,
  serviceAuth
};

File: 04_ai_engine_service/src/middleware/service_auth.py
"""
Service Authentication Middleware for AI Engine
JWT-based authentication for FastAPI endpoints
"""

import os
import jwt
import time
import logging
from datetime import datetime, timedelta
from typing import Optional, Dict, Any, List
from fastapi import HTTPException, Depends, Request
from fastapi.security import HTTPBearer, HTTPAuthorizationCredentials

logger = logging.getLogger(__name__)

class ServiceAuthenticator:
    def __init__(self):
        self.service_secret = os.getenv('SERVICE_JWT_SECRET')
        self.service_name = os.getenv('SERVICE_NAME', 'ai-engine')
        self.enabled = os.getenv('SERVICE_AUTH_ENABLED', 'false').lower() == 'true'
        self.valid_issuers = ['api-gateway', 'ai-engine', 'vision-service']
        
        if self.enabled and not self.service_secret:
            logger.warning("⚠️ SERVICE_JWT_SECRET not set but authentication is enabled")
        
        logger.info(f"🔐 Service Authentication: {'ENABLED' if self.enabled else 'DISABLED'}")

    def validate_service_token(self, token: str, expected_audience: Optional[str] = None) -> Dict[str, Any]:
        """Validate incoming service JWT token"""
        if not self.enabled:
            return {"valid": True, "bypass": True}
        
        if not token:
            return {
                "valid": False,
                "error": "No service token provided",
                "code": "MISSING_TOKEN"
            }
        
        if not self.service_secret:
            return {
                "valid": False,
                "error": "Service authentication not properly configured",
                "code": "AUTH_CONFIG_ERROR"
            }
        
        try:
            # Decode and verify JWT
            decoded = jwt.decode(
                token,
                self.service_secret,
                algorithms=['HS256'],
                options={"verify_exp": True}
            )
            
            # Validate audience if specified
            if expected_audience and decoded.get('aud') != expected_audience:
                return {
                    "valid": False,
                    "error": f"Invalid token audience. Expected: {expected_audience}, Got: {decoded.get('aud')}",
                    "code": "INVALID_AUDIENCE"
                }
            
            # Validate issuer
            issuer = decoded.get('iss')
            if issuer not in self.valid_issuers:
                return {
                    "valid": False,
                    "error": f"Unknown service issuer: {issuer}",
                    "code": "UNKNOWN_ISSUER"
                }
            
            return {
                "valid": True,
                "payload": decoded,
                "issuer": issuer,
                "audience": decoded.get('aud')
            }
            
        except jwt.ExpiredSignatureError:
            return {
                "valid": False,
                "error": "Service token has expired",
                "code": "TOKEN_EXPIRED"
            }
        except jwt.InvalidTokenError as e:
            return {
                "valid": False,
                "error": f"Invalid service token: {str(e)}",
                "code": "INVALID_TOKEN"
            }
        except Exception as e:
            logger.error(f"Token validation error: {e}")
            return {
                "valid": False,
                "error": "Token validation failed",
                "code": "VALIDATION_ERROR"
            }

    def extract_token(self, request: Request) -> Optional[str]:
        """Extract token from request headers"""
        # Check Authorization header
        auth_header = request.headers.get('authorization')
        if auth_header and auth_header.startswith('Bearer '):
            return auth_header[7:]
        
        # Check custom service header
        return request.headers.get('x-service-token')

    def create_dependency(self, expected_audience: Optional[str] = None):
        """Create FastAPI dependency for service authentication"""
        def authenticate_service(request: Request):
            if not self.enabled:
                return {"bypass": True, "service": "bypass"}
            
            token = self.extract_token(request)
            validation = self.validate_service_token(token, expected_audience)
            
            if not validation["valid"]:
                raise HTTPException(
                    status_code=401,
                    detail={
                        "error": "Service Authentication Failed",
                        "message": validation["error"],
                        "code": validation["code"],
                        "timestamp": datetime.utcnow().isoformat()
                    }
                )
            
            return {
                "valid": True,
                "issuer": validation.get("issuer"),
                "audience": validation.get("audience"),
                "payload": validation.get("payload", {})
            }
        
        return authenticate_service

    def get_status(self):
        """Get authentication status"""
        return {
            "enabled": self.enabled,
            "service_name": self.service_name,
            "has_secret": bool(self.service_secret),
            "valid_issuers": self.valid_issuers
        }


# Global authenticator instance
service_auth = ServiceAuthenticator()

# FastAPI dependencies
def require_service_auth(expected_audience: str = None):
    """Dependency that requires valid service authentication"""
    return Depends(service_auth.create_dependency(expected_audience))

def optional_service_auth():
    """Dependency that optionally validates service authentication"""
    def authenticate_optional(request: Request):
        if not service_auth.enabled:
            return {"bypass": True}
        
        token = service_auth.extract_token(request)
        if not token:
            return {"authenticated": False, "reason": "no_token"}
        
        validation = service_auth.validate_service_token(token)
        if validation["valid"]:
            return {
                "authenticated": True,
                "issuer": validation["issuer"],
                "payload": validation["payload"]
            }
        else:
            return {
                "authenticated": False,
                "reason": validation["code"],
                "error": validation["error"]
            }
    
    return Depends(authenticate_optional)


# Middleware for request logging with service info
async def service_auth_middleware(request: Request, call_next):
    """Middleware to add service authentication info to request"""
    start_time = time.time()
    
    # Extract and validate service token
    if service_auth.enabled:
        token = service_auth.extract_token(request)
        if token:
            validation = service_auth.validate_service_token(token)
            request.state.service_auth = validation
        else:
            request.state.service_auth = {"valid": False, "error": "No token provided"}
    else:
        request.state.service_auth = {"bypass": True}
    
    # Process request
    response = await call_next(request)
    
    # Add processing time and service info to response headers
    process_time = time.time() - start_time
    response.headers["X-Process-Time"] = str(process_time)
    
    if hasattr(request.state, 'service_auth') and request.state.service_auth.get('valid'):
        response.headers["X-Service-Issuer"] = request.state.service_auth.get('issuer', 'unknown')
    
    return response


# Health check with service authentication
def create_authenticated_health_check():
    """Create health check endpoint that requires service authentication"""
    def health_check_with_auth(service_info = require_service_auth()):
        return {
            "status": "healthy",
            "service": "ai-engine",
            "timestamp": datetime.utcnow().isoformat(),
            "authenticated": not service_info.get("bypass", False),
            "issuer": service_info.get("issuer") if not service_info.get("bypass") else None
        }
    
    return health_check_with_auth


# Export the main components
__all__ = [
    'ServiceAuthenticator',
    'service_auth',
    'require_service_auth',
    'optional_service_auth',
    'service_auth_middleware',
    'create_authenticated_health_check'
]